pinot-core/src/main/java/org/apache/pinot/core/query/reduce/StreamingReduceService.java

```diff
@@ -0,0 +1,114 @@
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.core.query.request.context.utils.BrokerRequestToQueryContextConverter;
+import org.apache.pinot.core.transport.ServerRoutingInstance;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The <code>StreamingReduceService</code> class provides service to reduce grpc response gathered from multiple servers
+ * to {@link BrokerResponseNative}.
+ */
+@ThreadSafe
+public class StreamingReduceService extends BaseReduceService {
+  private static final Logger LOGGER = LoggerFactory.getLogger(StreamingReduceService.class);
+
+  public StreamingReduceService(PinotConfiguration config) {
+    super(config);
+  }
+
+  public BrokerResponseNative reduceOnStreamResponse(BrokerRequest brokerRequest,
+      Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponseMap, long reduceTimeOutMs,
+      @Nullable BrokerMetrics brokerMetrics) {
+    if (serverResponseMap.isEmpty()) {
+      // Empty response.
+      return BrokerResponseNative.empty();
+    }
+
+    // prepare contextual info for reduce.
+    PinotQuery pinotQuery = brokerRequest.getPinotQuery();
+    Map<String, String> queryOptions =
+        pinotQuery != null ? pinotQuery.getQueryOptions() : brokerRequest.getQueryOptions();
+    boolean enableTrace =
+        queryOptions != null && Boolean.parseBoolean(queryOptions.get(CommonConstants.Broker.Request.TRACE));
+
+    QueryContext queryContext = BrokerRequestToQueryContextConverter.convert(brokerRequest);
+
+    String tableName = brokerRequest.getQuerySource().getTableName();
+    String rawTableName = TableNameBuilder.extractRawTableName(tableName);
+
+    // initialize empty response.
+    BrokerMetricsAggregator aggregator = new BrokerMetricsAggregator(enableTrace);
+
+    // Process server response.
+    DataTableReducerContext dataTableReducerContext =
+        new DataTableReducerContext(_reduceExecutorService, _maxReduceThreadsPerQuery, reduceTimeOutMs,
+            _groupByTrimThreshold);
+    StreamingReducer streamingReducer = ResultReducerFactory.getStreamingReducer(queryContext);
+
+    streamingReducer.init(dataTableReducerContext);
+
+    // TODO: use concurrent process instead of determine-order for-loop.
+    for (Map.Entry<ServerRoutingInstance, Iterator<Server.ServerResponse>> entry: serverResponseMap.entrySet()) {
+      Iterator<Server.ServerResponse> streamingResponses = entry.getValue();
+      while (streamingResponses.hasNext()) {
+        Server.ServerResponse streamingResponse = streamingResponses.next();
+        try {
+          DataTable dataTable = DataTableFactory.getDataTable(streamingResponse.getPayload().asReadOnlyByteBuffer());
+          streamingReducer.reduce(entry.getKey(), dataTable);
```

Comment:
Reducer should only reduce on data block, aggregator should only aggregate on metadata block, let's handle them separately